lib/segment/src/index/payload_config.rs

```diff
@@ -34,3 +34,106 @@ impl PayloadConfig {
         Ok(atomic_save_json(path, self)?)
     }
 }
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct PayloadFieldSchemaWithIndexType {
+    #[serde(flatten)]
+    pub schema: PayloadFieldSchema,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub index_types: Vec<FullPayloadIndexType>,
+}
+
+impl PayloadFieldSchemaWithIndexType {
+    pub fn new(schema: PayloadFieldSchema, index_types: Vec<FullPayloadIndexType>) -> Self {
+        Self {
+            schema,
+            index_types,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub enum FullPayloadIndexType {
+    IntIndex(IndexMutability),
+    DatetimeIndex(IndexMutability),
+    IntMapIndex(IndexMutability),
+    KeywordIndex(IndexMutability),
+    FloatIndex(IndexMutability),
+    GeoIndex(IndexMutability),
+    FullTextIndex(IndexMutability),
+    BoolIndex(IndexMutability),
+    UuidIndex(IndexMutability),
+    UuidMapIndex(IndexMutability),
+    NullIndex(IndexMutability),
+}
+
+impl FullPayloadIndexType {
+    pub fn mutability(&self) -> &IndexMutability {
+        match self {
+            FullPayloadIndexType::IntIndex(index_mutability)
+            | FullPayloadIndexType::DatetimeIndex(index_mutability)
+            | FullPayloadIndexType::IntMapIndex(index_mutability)
+            | FullPayloadIndexType::KeywordIndex(index_mutability)
+            | FullPayloadIndexType::FloatIndex(index_mutability)
+            | FullPayloadIndexType::GeoIndex(index_mutability)
+            | FullPayloadIndexType::FullTextIndex(index_mutability)
+            | FullPayloadIndexType::BoolIndex(index_mutability)
+            | FullPayloadIndexType::UuidIndex(index_mutability)
+            | FullPayloadIndexType::UuidMapIndex(index_mutability)
+            | FullPayloadIndexType::NullIndex(index_mutability) => index_mutability,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub enum IndexMutability {
+    Mutable(StorageType),
+    Immutable(StorageType),
+    Mmap { is_on_disk: bool },
+}
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub enum StorageType {
+    Gridstore,
+    RocksDB,
+    Mmap { is_on_disk: bool },
+}
```

Comment:
Note to others: the key types, defining payload index/storage variants, are here.